### ISSUE TITLE
choppy movement, freezes occasionally

### ISSUE THREAD

**Issue by ghost** [DatOneLefty/territory.cf#2]

**Comment by lholliger** [comment]
yep. im looking at a way to make the backend send a more compressed map of claims, it could also be that stuff is done synchronously for http requests, i need help with fixing that :/

**Comment by ghost** [comment]
Only send part of the map

**Comment by ghost** [comment]
maybe 1.5x screen area and keep fetching on every move

**Comment by lholliger** [comment]
it fetches the screen every move you do, i need to look into sending certain portions of the screen, the current fetch system sends it like every block is claimed (it returns the unclaimed color)

**Comment by ghost** [comment]
Is this running on an ssd?  You seem to be doing a lot of disk writes...

**Comment by ghost** [comment]
And canvas is probably faster than DOM? (or at least better for what you're doing)

**Comment by lholliger** [comment]
never thought of using canvas. The system is using an HDD, but i am planning to use a ramdisk that is constantly restoring to the HDD, because, if the system goes down, i dont think loosing like 5 blocks is a big issue

**Comment by lholliger** [comment]
i havent really learned canvas, soooo... ill attempt on a seperate page (probably territory.cf/canvas-beta.html)

**Comment by lholliger** [comment]
issue partially solved: the game now completely draws using canvas

the only issue is sync speed

**Comment by ghost** [comment]
maybe just sustain a socket and send the initial map ex:
```
xxxxxx   xxxx
 xx   xxxxx
xx   xx     xxxxxxx
                  x
```
then send messages of the form: `[color, x-coordinate, y-coordinate]`

and the client can do the computing

**Comment by lholliger** [comment]
how would i do sockets? i have only worked with synchronous get requests lol

**Comment by ghost** [comment]
Idk in php; Node makes it easier because you can handle requests in a more low-level way (and use socket.io)

**Comment by lholliger** [comment]
I think I'll use node for the map and movement system, but the initial game startup will use php to get the coordinates and other data values. I can't program at the moment so I'll do it sometime soon

**Comment by ghost** [comment]
I can PR

**Comment by ghost** [comment]
Ok I have the http server and stuff set up and I probably will need to rewrite most of territory.js lol

**Comment by ghost** [comment]
socket.io is cool

**Comment by lholliger** [comment]
Rewriting the engine would make it very hard for me to make modifications. Thanks for the help, but I don't know what I would do because I don't know much about node,

**Comment by ghost** [comment]
you'll see

**Comment by lholliger** [comment]
If you do a PR, I might use parts of the video but not all of it, I will definitely be switching to sockets, is it posible with python?

**Comment by ghost** [comment]
video?

**Comment by ghost** [comment]
Socket.io is cool because it is event-driven
ex on server:
```
io.on('connection', function(socket) {
  // socket is individual connection
  // when client does io.emit('hi', [data is optional])
  socket.on('hi', function(data) {
    // io.emit sends message to all connected clients
    io.emit('hello', socket.id)
  })
})
```

**Comment by lholliger** [comment]
the code i would need most is it to tell the server "Hey! im playing the game!" and whenever it asks for the map/moves itll return the info. Dont worry about rewriting the map reciever, i might just leave that up to Python or some function ill rewrite

**Comment by ghost** [comment]
But thats still kinda slower

**Comment by ghost** [comment]
I would have the server do `io.emit('+', {x: [x], y: [y]})` so the client can do it and its only a few bytes being sent on a move (but send the whole map when starting the game)

**Comment by lholliger** [comment]
wouldn't that send it to all the players?

**Comment by ghost** [comment]
The client can only send to the server (for security reasons and technical ones)

**Comment by ghost** [comment]
So the client would say "hey server I wanna move here"
"ok mr. dr. sir"
"everyone, in your records make [x, y] [color]"
"Ok server"
"Ok server"

**Comment by lholliger** [comment]
im going to mess with socket.io a little for myself to try some things :)

**Comment by ghost** [comment]
ok ;) but I already have some stuff done ok

**Comment by lholliger** [comment]
waaiiittt. wat port will i need to open on mai server

**Comment by ghost** [comment]
You can choose.
if linux/mac, do export PORT=[port]

**Comment by lholliger** [comment]
im just wondering, why are you rewriting basiclly the entire game? it kinda seems like you are taking over almost all of it

**Comment by ghost** [comment]
If you used the express boilerplate, edit bin/www and change 3000 to whatever (put using the command thing will override)

**Comment by ghost** [comment]
Because it's fun and I need practice lol

**Comment by lholliger** [comment]
i just used npm install socket.io soooo

**Comment by ghost** [comment]
I never can come up with ideas and I was bored and this looked kinda sad (no offence)... So i decided... Why not rewrite something existing...

**Comment by ghost** [comment]
https://socket.io/get-started/chat/ look through this

**Comment by lholliger** [comment]
im kinda loosing control of the project it seems...

**Comment by ghost** [comment]
How?  It's just a fork, and I can't modify this repo :P

**Comment by lholliger** [comment]
i know, ill probably use some but not all of the code, ill probably add some changes to it too when you publish it

**Comment by lholliger** [comment]
on another thing, i think im going to leave PHP up to the inital map and variable sending when you first join the game, and then let socket.io take over most of it after that

**Comment by ghost** [comment]
Kind of weird but ok...

**Comment by lholliger** [comment]
it doenst make it so we have to rewrite a bunch so it saves time

**Comment by ghost** [comment]
But wouldn't it make more sense to just hold everything in a multidimensional array anyways?

**Comment by ghost** [comment]
oh well

**Comment by lholliger** [comment]
So so far what do you have written

**Comment by ghost** [comment]
Well the web server, have socket.io up, and am working on a mechanism to send the map

**Comment by lholliger** [comment]
im getting a little better at it. looked at some of your code, took some code for returning resources (css), and currently it returns the coords of whatever i tell the server to when you start. now to have it restore sessions...

**Comment by ghost** [comment]
ok cool

**Comment by lholliger** [comment]
currently i have it to save sessions... now... for... the....impossible...part..... maps

**Comment by ghost** [comment]
Can you put the code in a new branch?

**Comment by ghost** [comment]
please

**Comment by lholliger** [comment]
im putting it in a seperate project lol

**Comment by ghost** [comment]
Why?

**Comment by lholliger** [comment]
becuase im not good at git and im lazy to find out... rip

**Comment by ghost** [comment]
lol just use github desktop or gitkraken then

**Comment by lholliger** [comment]
im using ubuntu from my server controlling it from a mac. i dont have a working laptop anymore :(

**Comment by ghost** [comment]
But macs are gud for dev

**Comment by lholliger** [comment]
but is from 2007 and is having some truble running TeamViewer, Spotify, and Chome, and i dont have any good editors installed

**Comment by lholliger** [comment]
wut if as a prank i put all the instructions (when i add them) in leetspeak pro mode

**Comment by lholliger** [comment]
how can I make my code (line 41-65 in https://github.com/DatOneLefty/territory.cf-socket) get the file contents before returning it?

**Comment by lholliger** [comment]
and sending the map would be an issue especially when every time you move it needs a new chunk, and sending the entire map at once may not work well

**Comment by ghost** [comment]
Just send everyone
"hey in your cache of the map change [x, y] to [color]"

**Comment by lholliger** [comment]
my current system just adds it to a list and removes the first duplicate, but its breaking stuff i think, i think to fix it temporarly, i need to request a new map every few steps, at this point, it isnt that much

**Comment by lholliger** [comment]
issue fixed with beta 2.0 replacing php with node
